### ISSUE TITLE
Downloading Setup Script: ERROR 404 Not Found

### ISSUE THREAD

**Issue by elJoeyJojo** [kireevco/wimaging#30]
Hello,

I am having a problem with WinPE that keeps trying to look for an /unattended/script file but it doesn't find it.

![image](https://cloud.githubusercontent.com/assets/17091968/12995766/c94b3dc6-d0de-11e5-8fc5-537c130c5b3a.png)

The templates WAIK peSetup, WAIK provision unattended.xml, WAIK PXE Linux, WAIK finish, WAIK local users, WAIK extra finish commands all came from this the git master wimaging. They all render without any issue reporting but for some reason it always craps out at the "Downloading Setup Script" 
WinPE loads correctly, starts Networking, then it complains the subsystem needed to support the image type is not present. Then "ping to puppet.domain.com OK! Your IP Address. And then it errors out at the "Downloading Setup Script" 

Can anybody help out? Also can anybody explain what part of the template is not working? Would I be correct to assume that the template that is not working is the WAIK peSetup and if so, that link" http://puppet.domain.com/unattended/script: 
Where is that in the code for the WAIK peSetup template? Please help.

Thank You

**Comment by DimmKirr** [comment]
@elJoeyJojo Are you sure your request comes from the same IP address that foreman gave it?
The way foreman authenticates the request is by IP.

This will be an issue if you are running your provisionning behind NAT.

**Comment by elJoeyJojo** [comment]
@kireevco you're a genius sir! it was not the same IP so I changed it and VOILA! thanks Kireevco! Now to the next issue lol 

Error: 50

DISM does not support servicing a Windows Vista RTM or earlier os. 
If the os is supported check that  SSShim.dll is present.
The DISM log file can be found at x:\windows\Logs\DISM\dism.log
Applying Driver

Deployment Image Servicing and Management Tool
Version 6.1.7600.16385

Thanks for getting me this far though Kireevco

**Comment by DimmKirr** [comment]
@elJoeyJojo that probably means that your global variables are not set. Specifically it feels like that OS edition is not set. Expected variables are listed in the templates. I've udpated templates a little bit, verified they all work as expected. so do something like that on your foreman host:
(Just make sure to keep your local changes safe somewhere)

``` bash
foreman-rake apipie:cache
foreman-rake templates:sync
foreman-rake templates:sync repo="https://github.com/kireevco/foreman-wimaging.git" branch='master' verbose='true'
```

**Comment by elJoeyJojo** [comment]
hey @kireevco Thanks for being so prompt as always. I actually saw that it might be something else as I get a "Downloading User Data" it resolves to puppet, and then I get the 404 Not Found. 
Then it creates a "temp" staging folder for DISM and then it gives me the Error: 50 

![image](https://cloud.githubusercontent.com/assets/17091968/13060134/1b78e88e-d3e4-11e5-90d1-85171d11cb3d.png)

**Comment by DimmKirr** [comment]
How about dns resolution? In pecmd template enable debug (uncomment DEBUG at the end of the script) and run diagnostics like Nslookup and wget64

**Comment by elJoeyJojo** [comment]
@kireevco hey Dmitry - I can't seem to run nslookup (unrecognized)
I looked further into it and it craps out at the "Downloading user data" and I realized that I didn't have those WAIK UserDat_JoinDomain or the WAIK OU fromHostGroup so I added them in there. It be easier to just test without all these scripts and just use the necessary ones so can I just comment out the "user data" stuff and any other ones I don't need like JoinDomain, User Data and ExtraFinish Commands? They render without an issue but still get the same issue. There is a log that it creates in WinPe and this is what I was able to capture as I can't copy and paste from VM:

![l_8d28 tmp](https://cloud.githubusercontent.com/assets/17091968/13064522/a44d3e98-d408-11e5-8f23-97ea0c9315b5.PNG)
![l_5d53 tmp](https://cloud.githubusercontent.com/assets/17091968/13064521/a447ebe6-d408-11e5-9296-6783111770c6.PNG)

**Comment by DimmKirr** [comment]
Check if you can actually download your install.wim file.
What is your HTTP server directory structure?

**Comment by elJoeyJojo** [comment]
my ftp structure is http://server/pub/windows-pe-x64/sources/install.wim
I went to it via a browser and i downloaded the install.wim file in less than 2 minutes so that's all good there.

**Comment by DimmKirr** [comment]
@elJoeyJojo Just to confirm, did you set your global variable, specifically `wimImageName`? Are you sure it is correct and your install.wim contains such image name? (like `Windows Server 2012 R2 SERVERSTANDARD`)?

**Comment by elJoeyJojo** [comment]
@kireevco hey Dmitry, i fixed that Downloading User Data, I realized that it was asking for a template "WAIK user Data Domain Join that I hadn't created 'cause I didn't wanna use it. I added it and it no longer gives me the Error 404 Not Found but I do still have the 
Error 50 
DISM does not support servicing a Windows Vista RTM or earlier OS

I did set the global variable

![image](https://cloud.githubusercontent.com/assets/17091968/13085362/c357362e-d492-11e5-9732-5aa7973080b1.png)

However looking at my /inc/config\* files I noticed that both the:

win81x64-Config and winpe-x64-Config are both set to:

$edition = " Windows 8.1 Enterprise"

**Comment by DimmKirr** [comment]
So try to change it to `Windows 8.1 Enterprise`?

**Comment by elJoeyJojo** [comment]
@kireevco it didn't work so I will just re-create the images from the beginning and see if that fixes anything

**Comment by elJoeyJojo** [comment]
@kireevco hey so I redid everything from beginning and for some reason my Foreman cannot get to my FTP server: nagios.domain.com
I am able to get to http://nagios.domain.com/pub/windows81x64/sources/install.wim but for some reason it gets a "failed" as you can see in the picture. I checked from another PC to see if it was prompting username/password to get to http://nagios.domain.com/pub but it has no security. Am I missing something? I mean it should be able to get the install.wim file without any issue but it craps out.

![image](https://cloud.githubusercontent.com/assets/17091968/13306415/36d4871a-db18-11e5-8223-45fbcc4f2678.png)

This picture shows you what my FTP server looks like:

![image](https://cloud.githubusercontent.com/assets/17091968/13306505/1d1b42ae-db19-11e5-9a12-5249329358aa.png)

**Comment by elJoeyJojo** [comment]
@kireevco hey sorry i didn't get back to you for a while I was taking a break from Foreman. 
Ok so re-did everything and it fails with mostly with anything that runs dism.exe

Mine errors out after:

"echo Write the Install image to the partition"
dism.exe /apply-image /imagefile:c:\install.wim /name:"Windows 8.1 Pro" /ApplyDir:C:\

![image](https://cloud.githubusercontent.com/assets/17091968/13306810/98218272-db1b-11e5-9c5a-a7e2e729483d.png)

Any idea ?

**Comment by elJoeyJojo** [comment]
@kireevco I found this on the net, https://social.msdn.microsoft.com/Forums/en-US/1ea3c5fc-f0e2-44ff-881f-0312e9649e29/getting-dism-error-87-the-applyimage-option-is-unknown?forum=WindowsIoT

Did you use Windows 10 ADK?

**Comment by elJoeyJojo** [comment]
@kireevco I am confused with these instructions Dmitry sorry but they are not clear at all and every step I make is almost backwards. For example when the Foreman server has a boot folder in /var/lib/tftpboot/boot and in there it has a windows_x64 folder that has the following files:

bcd, bootmgr, boot.sdi, boot.wim, install.wim wimboot

Where did these come from? I do not remember adding these in here and even if I don't, the actual MAC 01-00-00-58-XX-XX-XX file has this command:

DEFAULT wimbootpe
LABEL wimbootpe
     com32 linux.c32 boot/wimboot
     append initrdfile=boot/windows-x64/bcd,boot/windows-x64/boot.sdi,boot/windows-x64/boot.wim

I am confused because I have no idea why these files are being pulled locally when they are on my FTP server? If I change the FTP files (like I did already) Do these eventually sync?

I wish someone could write a Step by Step instructions on how to do it as this point form instructions lacks a lot of detail that is causing delays and delays. My previous issues do not like certain dism.exe commands so I thought I'd use a newer version of WinPE and then I replaced the boot.wim file in /var/tftpboot/boot/windows-x64/ with a new version of WinPE and now it craps out here without doing anything:

![image](https://cloud.githubusercontent.com/assets/17091968/13333785/97a33884-dbbd-11e5-8ad4-262be4e792a8.png)

**Comment by elJoeyJojo** [comment]
I give up on this Foreman Windows Imaging! The instructions are all over the place and every step I make to fix it only breaks it furthermore. Now I can't even get WinPE to run any of the scripts. The documentation is just not ready. I have no clue what WinPE (people who set it up successfully) to use, why the /var/lib/tftboot/boot/windows-x64 folder has the files that are suppose to come from the http://ftp.server.com/pub/ 

I tried to use a Windows 8.1 ADK to create the boot.wim (winpe) and copied it over to the http://ftp.server.com and re-tried but same issue. So I thought ok, I can replace that boot.wim file in /var/lib/tftpboot/boot/windows-x64 with the new version and the look of WinPE changed but now I can't even run any commands. I have no idea where to go, I'm sure you are very aware of what you did but with no real instructions detailing the steps it creates a lot of holes and just keep going in loops. 

Hopefully one day someone documents it in full detail, with versions, folder structure, explanations to why and where etc...

WDS will have to do for now.

**Comment by ghost** [comment]
I'm on the same boat as you, elJoeyJojo.
Struggling to get past the wpeinit step. In wpeinit.log it only says, it couldn't find an unattend xml file.

I really hope, documentation will be updated to contain much more details so we can finally use this great project for our environments.

**Comment by thatguydev** [comment]
After two weeks of working on this, I finally got past the provisioning state. I was able to do a domain join BUT left puppet out as I was trying to keep this as skinny as possible to test the concept. I ran into most of the problems as described here, some are due to "light" documentation and some are plain old "user" mistakes :) 

To answer @elJoeyJojo /@St3ve87 current state. If the window is similar to the screenshot that @elJoeyJojo posted, [x:\windows\system32>], that means your PE wasn't built correctly. It should be [x:\wimaging\deploy]. Check `ls -la /var/lib/tftpboot/boot/windows-x64/` and see if all the files downloaded correctly. The directory should include [ bcd | bootmgr | boot.sdi | boot.wim | wimboot ]

In saying all that, these were the steps taken by me to get to the state where I am now, I did it twice: 

Preface: 
- Fresh (un-managed) 8.1 T430s laptop with all the updates
- Used NotePad ++ vs notepad  
- Copied 8.1 Prof to a folder[8.1Disk] on my desk
- Fresh Ubuntu 14.04 (foreman- dhcp -tftpboot) laptop
- VMware with one DC (DNS) and one ubuntu provisioned by Foreman, to verify that my install of foreman is solid. 
- Updated to SysLinux 5.01
- Made sure no previous [macAddress] existed in /var/lib/tftpboot/pxelinux.cfg
- Rm -rf /var/lib/tftboot/boot/Windows-x64
- IIS Setup: Enable directory reading | Add MIME Extension for “*”, all\all (devEnvironment)

Side Note: I didn’t bother w/drivers or change(s) in template(s). Everything is stock except for joinDomain.ps1. I removed the ‘OUPath’ :) These steps aren't meant to replace the current instructions, this is merely my workflow for illustration purposes. 

My workflow: 
**Step 1**: Download the [wimaging-master] directory and the [Tools] directory
    Note: [Tools] never was compiled in the [Wimaging-master] archive
**Step 2**: Right click each zip and “unblock” all files
**Step 3**: No changes in “[os]-Config.ps1”
**Step 4**: (II. Configure wimaging) Tools: 
Add Foreman IP
Rename to “CMD”
**Step 5**: Copy and rename [install][directory.template] → [win81x64]
**Step 6**: 
Copied **ALL** boot contents from [8.1Disk] to wimaging/sources/win81x64/boot
Copied install.wim from [8.1Disk] to  wimaging/sources/win81x64/sources
Copied bootmgr/bootmgr.efi from [8.1Disk] to  wimaging/sources/win81x64/
**Step 7**: Download wimboot and place the file to  wimaging/sources/win81x64/ (this wasn't mentioned anywhere)
**Step 8**: Place the winpe.wim from [Windows 8.1\Assessment and Deployment\Windows Preinstallation Environment\amd64] to wimaging/sources/windows-pe-x64/
Note make sure you get the 64bit
**Step 9**: Use the [Deployment & Imaging Tools Environment] cmd as Admin and run boot.wim/install.wim procedure**
**Step 10**:  (the files listed below should have been copied over from step 6 to [install] but they didn’t, so I manually placed them) 
Copied [bcd | boot.sd]  boot contents from [8.1Disk] to wimaging/install/win81x64/boot
Copied [bootmgr|bootmgr.efi] root contents from [8.1Disk] to wimaging/install/win81x64/
Copied [wimboot] file to wimaging/install/win81x64/
**Step 11**: 
Rebuild Config and make sure templates render
**Step 12**:
Check if files are correctly downloaded (make sure they have valid a size and not 0) post config bulild: 
`$ ls -la /var/lib/tftboot/boot/windows-x64/`

btw @kireevco brilliant implementation, thanks for your hard work. Looking forward to contributing.

**Comment by elJoeyJojo** [comment]
@thatguydev hey dude! thanks for giving us your instructions. I just have a few questions. Step 8. you mention to place the wipe.wim from (Windows...Environment\amd64) where is that path coming from?

Do you mean from the Windows 8.1\Assessment and Deployment\Windows Preinstallation Environment\amd64\ ?

**Comment by thatguydev** [comment]
@elJoeyJojo - That is correct. Added the change.

**Comment by elJoeyJojo** [comment]
@thatguydev in the original documentation III. Prepare install.wim and boot.wim. I can run ./Init-WorkWim.ps1. just fine but when I run .\Update-All.ps1 I get all there errors! 
Error 2

Unable to access the image.

I am following every step you did. By the way Step 9 doesn't make sense to me. You say "Use Deployment & Imaging Tools Environment" and run boot.wim/install.wim procedure but in that procedure you have to run "run-wimaging-shell.cmd" which is in c:\wimaging\ so how does that work? If I open the Deployment & Imaging Tools Environment there is now way I can run the \Init-WorkWim.ps1

**Comment by elJoeyJojo** [comment]
@thatguydev although after all those errors I do get at the bottom: 
Scanning drive C for stale files
The operation completed successfully.
(in RED) SafeUnmountWim : Wim was not pushed 

Did you get this at all or did your go through without any errors?

**Comment by elJoeyJojo** [comment]
@thatguydev ok this is a stupid question that I didn't have before but now I am getting a 
"The disk you specified is not valid" 
c:/install.wim: No such file or directory. 

I want to know how the c: partition is created?

**Comment by elJoeyJojo** [comment]
@thatguydev ignore that last one, seemed to be a disk issue and i deleted the VM and recreated it and then it recognized that c:disk 
I am now at the same point i once was with the error:

Error: 50

DISM does not support servicing a Windows Vista RTM.

Basically the last few steps of the WAIK peSetup Template do not run:

echo Apply Unattend.xml
dism.exe /Image:C:\ /Apply-Unattend:c:\Windows\Panther\unattend.xml /ScratchDir:C:\MININT\Scratch/

echo Apply Drivers
dism.exe /Image:C:\ /Add-Driver /Driver:c:\drivers\ /Recurse /ForceUnsigned

Any idea? Anyone?

**Comment by thatguydev** [comment]
[Q] Prepare install.wim and boot.wim. I can run ./Init-WorkWim.ps1. just fine but when I run .\Update-All.ps1 I get all there errors! 
Error 2

[A] Are you running this as Admin ? 

[Q]I am following every step you did. By the way Step 9 doesn't make sense to me. You say "Use Deployment & Imaging Tools Environment" and run boot.wim/install.wim procedure but in that procedure you have to run "run-wimaging-shell.cmd" which is in c:\wimaging\ so how does that work? If I open the Deployment & Imaging Tools Environment there is now way I can run the \Init-WorkWim.ps1

[A] Step 9, Deployment & Imaging Tools Environment is part of the WAIK, so it starts it's environment at that location.  You have to "Cd C:[WimagingPath] then run-wimaging-shell.cmd

[Q] Did you get this at all or did your go through without any errors?
[A] Once I used fresh tech machine, I didn't receive any errors

[Q] I assume this is the part of the script that assigns a partition but then why am I getting those "disk not valid" errors?

[A] Yes, it creates and assigns the partition. Terminate the batch script and run it manually to pin point the error. Run each command as it is displayed in [wimaging_partition_table].

**Comment by elJoeyJojo** [comment]
@thatguydev You said you used a fresh tech machine. Mmmm I've been running this on my normal PC  but I am not sure why this even matters. When you say you used a fresh tech machine, do you mean you just setup a Win 8.1 VM and do all the work from there right? Downloading the master git zip file, install WAIK and run all the commands from within that VM right? I haven't tried that yet as I thought it didn't really matter

**Comment by elJoeyJojo** [comment]
@thatguydev at the end of running all the configuration (powershell scripts) i get a directory as such:
c:\wimaging\install\win81x64\boot, drivers, extras, sources (folders) and bootmgr, bootmgr.efi, wimboot (files)
c:\wimaging\install\windows-pe-x64\EMPTY (shouldn't there be a file here)???

**Comment by elJoeyJojo** [comment]
@kireevco @thatguydev  no, i absolutely give up on this project! i have spent way too many hours trying to get a simple installation of Windows 8.1 and the plan was to go fwd and setup all Windows Servers using Foreman but I just cannot. 

I followed @thatguydev 's steps and it didn't work so I thought ok, I will set this brand new Windows 8.1 VM and download the master, copy the files from one folder to the other etc.... then uploaded it to my ftp server and confirmed Foreman fetched those 2 folders from my FTP server. I even removed that whole folder in /var/lib/tftpboot/boot/windows-x64 ... 

Now I can't even start the first command and I get this! 

The syntax of the command is incorrect

x:\wimaging\deploy> 

Thanks for trying to help but there are too many questions and nobody has the patience and time to document inside out how they got it to work. Good luck to the rest who keep trying but as for me, I'm gonna stop 'cause I'm starting to hate Foreman and I used to love it.
